File: java/src/main/java/deltix/luminary/CompositeTypeScope.java
package deltix.luminary;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class CompositeTypeScope<TProperty, TType extends Type> extends CompositeDef<TProperty> implements TypeScope, TypeDef<TType> {
    private final FileDef file;
    private final TypeScope parent;
    private List<ClassDef> definedClasses = new ArrayList<>();
    private List<EnumerationDef> definedEnumerations = new ArrayList<>();
    private List<InterfaceDef> definedInterfaces = new ArrayList<>();
    private List<DecoratorDef> definedDecorators = new ArrayList<>();

    protected CompositeTypeScope(@NotNull TypeScope parent, @NotNull String name, @Nullable List<String> comments) {
        super(name, parent instanceof FileDef ? name : ((ItemDef) parent).getFullName() + "." + name, comments);
        this.parent = parent;
        this.file = parent.getFile();
    }

    @Override
    @NotNull
    public abstract TType getType();

    @Override
    @NotNull
    public List<ClassDef> getDefinedClasses() {
        return definedClasses;
    }

    @Override
    @NotNull
    public List<EnumerationDef> getDefinedEnumerations() {
        return definedEnumerations;
    }

    @Override
    @NotNull
    public List<InterfaceDef> getDefinedInterfaces() {
        return definedInterfaces;
    }

    @Override
    @NotNull
    public List<DecoratorDef> getDefinedDecorators() {
        return definedDecorators;
    }

    @Override
    @NotNull
    public TypeScope getParent() {
        return parent;
    }

    @Override
    @NotNull
    public FileDef getFile() {
        return file;
    }
}

File: csharp/main/CompositeTypeScope.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;

namespace Deltix.Luminary
{
	public abstract class CompositeTypeScope<TProperty, TType> : CompositeTypeDef<TProperty>, ITypeScope, ITypeDef<TType>
		where TType : Type
	{
		protected CompositeTypeScope([NotNull] String name, [NotNull] ITypeScope parent, [CanBeNull, ItemNotNull] List<String> comments)
			: base(name, comments)
		{
			Parent = parent;
			File = parent.File;

			DefinedClasses = new List<ClassDef>();
			DefinedEnumerations = new List<EnumerationDef>();
			DefinedInterfaces = new List<InterfaceDef>();
			DefinedDecorators = new List<DecoratorDef>();
		}

		public List<ClassDef> DefinedClasses { get; }

		public List<InterfaceDef> DefinedInterfaces { get; }

		public List<EnumerationDef> DefinedEnumerations { get; }

		public List<DecoratorDef> DefinedDecorators { get; }

		[NotNull]
		public FileDef File { get; }

		[NotNull]
		public ITypeScope Parent { get; }

		[NotNull]
		public abstract TType Type { get; }
	}
}
